Remotion/Validation/Core/Validators/NotEmptyCollectionValidator.cs

```diff
@@ -0,0 +1,81 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// The re-motion Core Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+//
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using JetBrains.Annotations;
+using Remotion.FunctionalProgramming;
+using Remotion.Reflection;
+using Remotion.Utilities;
+using Remotion.Validation.Implementation;
+using Remotion.Validation.Results;
+
+namespace Remotion.Validation.Validators
+{
+  public class NotEmptyCollectionValidator : IRequiredValidator
+  {
+    public string ErrorMessage { get; }
+    public ValidationMessage ValidationMessage { get; }
+
+    public NotEmptyCollectionValidator ([NotNull] ValidationMessage validationMessage)
+    {
+      ArgumentUtility.CheckNotNull("validationMessage", validationMessage);
+
+      ErrorMessage = "The value must not be empty.";
+      ValidationMessage = validationMessage;
+    }
+
+    public IEnumerable<PropertyValidationFailure> Validate (PropertyValidatorContext context)
+    {
+      if (IsValid(context))
+        return Enumerable.Empty<PropertyValidationFailure>();
+
+      return EnumerableUtility.Singleton(CreateValidationError(context));
+    }
+
+    private bool IsValid (PropertyValidatorContext context)
+    {
+      var propertyValue = context.PropertyValue;
+
+      if (propertyValue == null)
+        return true;
+
+      if (propertyValue is ICollection collection)
+        return collection.Count > 0;
+
+      if (propertyValue is IReadOnlyCollection<object> readOnlyCollection)
+        return readOnlyCollection.Count > 0;
+
+      if (propertyValue.GetType().CanAscribeTo(typeof(ICollection<>)))
+        return ((IEnumerable<object>)propertyValue).Any();
+
+      throw new NotSupportedException($"Only the following types are supported: IReadOnlyCollection<T>, ICollection<T>, and ICollection.");
```

Comment:
The original implementation simply skipped incorrectly typed properties as valid. This behavior is true for all validators and should be preserved.